beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -441,6 +441,35 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             .map(|result| result.map_err(|e| e.into())))
     }
 
+    /// Iterate through the current chain to find the slot intersecting with the given beacon state.
+    /// The maximum depth this will search is `SLOTS_PER_HISTORICAL_ROOT`, and if that depth is reached
+    /// and no intersection is found, the finalized slot will be returned.
+    pub fn find_reorg_slot(
+        &self,
+        new_state: &BeaconState<T::EthSpec>,
+        new_block_root: &Hash256,
+    ) -> Result<Slot, Error> {
+        process_results(self.rev_iter_block_roots()?, |iter| {
```

Comment:
This call to `rev_iter_block_roots` will clone the head state, which is a pretty expensive operation.

The alternative to this clone would be to create a custom common-descendant-search using `BeaconChain::with_head`. This has the downside of holding a read-lock on `beacon_chain.head` for the duration of the search.

However, the duration of the search will be quite short (at most 8,096 `Hash256` equality checks). Additionally, only functions that want a *write*-lock on the head will be affected by this and AFAIK the only function that wants a write-lock is `BeaconChain::find_head` (this function!).

We call `fn find_head` fairly frequently during sync (IIRC, every 128 blocks). IIRC a state clone takes ~30ms, which would add 4 mins to syncing the entire chain at this point. That's not really a big deal, but considering the impact that state clones can have on memory fragmentation, I'd be tempted to optimize out the clone.

Reply:
refactored this, let me know what you think